### PR TITLE
Observed data cannot be logged

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

One of the primary motivations for the Observer interface is the
ability to log samples as they are written, received, and accessed.
Currently, users of the Observer interface receive a generic pointer
to sample data which may be a raw sample or something else.
There is no capability that allows a user to take this generic
pointer and log the sample contained therein.

Solution
--------

1. Standardize on the approach that the pointer in the Sample is a
pointer to a raw sample.

2. Introduce the ValueWriter interface and necessary support in the
IDL compiler to allow generic processing of the raw sample.

3. Add plumbing between generic and typed DataReader and DataWriter to
allow a user to dispatch a ValueWriter on the raw sample.

4. Provide an incomplete JsonValueWriter to demonstrate how types
processed by the IDL compiler could be serialized to JSON.

Notes
-----

* It may be necessary/beneficial to augment the ValueWriter interface
  with type information via TypeObjects.

* ValueWriter suggests a symmetric interface called ValueReader that
  takes an opaque representation and populates a raw sample, e.g.,
  deserialization.

### CODE REVIEW

File: dds/DCPS/DataReaderImpl_T.h [hunk 1]
```diff
@@ -30,8 +30,9 @@ namespace OpenDDS {
     OpenDDS_Dcps_Export
 #endif
     DataReaderImpl_T
-    : public virtual OpenDDS::DCPS::LocalObject<typename DDSTraits<MessageType>::DataReaderType>,
-      public virtual OpenDDS::DCPS::DataReaderImpl
+    : public virtual OpenDDS::DCPS::LocalObject<typename DDSTraits<MessageType>::DataReaderType>
+    , public virtual OpenDDS::DCPS::DataReaderImpl
+    , public OpenDDS::DCPS::ValueWriterDispatcher
```
Comment (mitza-oci) [inline dds/DCPS/DataReaderImpl_T.h:35]:
```suggestion
    : public virtual LocalObject<typename DDSTraits<MessageType>::DataReaderType>
    , public virtual DataReaderImpl
    , public ValueWriterDispatcher
```

File: dds/DCPS/DataReaderImpl_T.h [hunk 2]
```diff
@@ -250,9 +260,10 @@ namespace OpenDDS {
           ptr->instance_state_->sample_info(sample_info, item);
           item->sample_state_ = DDS::READ_SAMPLE_STATE;
 
-          if (observer) {
-            Observer::Sample s(*item, sample_info.instance_handle, sample_info.instance_state);
-            observer->on_sample_read(this, s);
+          const ValueWriterDispatcher* vwd = get_value_writer_dispatcher();
+          if (observer && item->registered_data_ && vwd) {
+            Observer::Sample s(sample_info.instance_handle, sample_info.instance_state, item->source_timestamp_, item->sequence_, item->registered_data_);
```
Comment (mitza-oci) [inline dds/DCPS/DataReaderImpl_T.h:265]: should there be a constructor that takes a `ReceivedDataElement` and gets these 3 parts out of it?

File: dds/DCPS/DataWriterImpl_T.h [hunk 3]
```diff
@@ -26,8 +26,9 @@ class
   OpenDDS_Dcps_Export
 #endif
 DataWriterImpl_T
-: public virtual OpenDDS::DCPS::LocalObject<typename DDSTraits<MessageType>::DataWriterType>,
-  public virtual OpenDDS::DCPS::DataWriterImpl
+: public virtual OpenDDS::DCPS::LocalObject<typename DDSTraits<MessageType>::DataWriterType>
+, public virtual OpenDDS::DCPS::DataWriterImpl
+, public OpenDDS::DCPS::ValueWriterDispatcher
```
Comment (mitza-oci) [inline dds/DCPS/DataWriterImpl_T.h:31]:
```suggestion
: public virtual LocalObject<typename DDSTraits<MessageType>::DataWriterType>
, public virtual DataWriterImpl
, public ValueWriterDispatcher
```

File: dds/idl/value_writer_generator.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    } else if (c & CL_ARRAY) {
+      AST_Array* array = dynamic_cast<AST_Array*>(actual);
+      array_helper(expression, array, 0, idx);
+    } else if (c & CL_FIXED) {
+      be_global->impl_ << "value_writer.write_fixed(" << expression << ");\n";
+    } else if (c & CL_STRING) {
+      if (c & CL_WIDE) {
+        be_global->impl_ << "value_writer.write_wstring(" << expression << ");\n";
+      }
+      else {
```
Comment (mitza-oci) [inline dds/idl/value_writer_generator.cpp:64]:
```suggestion
      } else {
```

File: dds/idl/value_writer_generator.cpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+      case AST_PredefinedType::PT_octet:
+        be_global->impl_ << "value_writer.write_byte(" << expression << ");\n";
+        break;
+      case AST_PredefinedType::PT_any:
+      case AST_PredefinedType::PT_object:
+      case AST_PredefinedType::PT_value:
+      case AST_PredefinedType::PT_abstract:
+      case AST_PredefinedType::PT_void:
+      case AST_PredefinedType::PT_pseudo:
+        // TODO
```
Comment (mitza-oci) [inline dds/idl/value_writer_generator.cpp:114]:
```suggestion
```
seems like we won't ever have/need support for these

File: dds/idl/value_writer_generator.cpp [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+  {
+    NamespaceGuard guard;
+
+    Function write("vwrite", "void");
+    write.addArg("value_writer", "OpenDDS::DCPS::ValueWriter&");
+    write.addArg("value", "const " + type_name + "&");
+    write.endArgs();
+
+    be_global->impl_ << "switch(value) {\n";
```
Comment (mitza-oci) [inline dds/idl/value_writer_generator.cpp:158]:
```suggestion
    be_global->impl_ << "switch (value) {\n";
```

File: dds/idl/value_writer_generator.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    be_global->impl_ << "}\n";
+  }
+
+  return true;
+}
+
+bool value_writer_generator::gen_typedef(AST_Typedef*,
+                                   UTL_ScopedName*,
+                                   AST_Type*,
+                                   const char*)
```
Comment (mitza-oci) [inline dds/idl/value_writer_generator.cpp:175]: reformat

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/DataReaderImpl_T.h (hunk 2 above)
Comment (jrw972) [reply dds/DCPS/DataReaderImpl_T.h:265]: I could go either way so I'll go ahead an add one.

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**